### ISSUE TITLE
Add examples or high-level documentation for Monix/cats-based implementation.

### ISSUE THREAD

**Issue by SergeantSod** [outwatch/outwatch#183]
I'm currently evaluating outwatch for a small project I'm planning. While initial experiments on an older version seemed really promising, I thought I'd give 1.0.0-RC2 a try, and some of my ideas don't really seem to be practical anymore with this version. Some small but complete examples would be helpful to figure out how to structure your app.

It would also help a lot if there was at least a high-level intro to some of the design choices. 

I'm specifically curious why `Handler.create[T]` returns an `IO`. It's not really obvious to me what the side-effect that we're wrapping in the `IO` here is. 

I'm particularly interested about this design change since it seems to make some of the design ideas I had less straight-forward.

**Comment by cornerman** [comment]
Thanks for opening an issue. We are currently working on a changelog to better explain the changes and ideas behind it (https://github.com/OutWatch/outwatch/pull/167). I agree, examples are still missing.

> I'm specifically curious why Handler.create[T] returns an IO. It's not really obvious to me what the side-effect that we're wrapping in the IO here is.

This is effectively a factory for an object that has sideeffecting functions like subscribe/onNext/onComplete, which is why it is modeled as an effect.

For IO, there is a work in progress to abstract over the cats.Effect type, which might be of interest: https://github.com/OutWatch/outwatch/pull/178. This also enables working with different effect types.

**Comment by fdietze** [comment]
> some of my ideas don't really seem to be practical anymore with this version

> it seems to make some of the design ideas I had less straight-forward.

Which were your ideas? Can you show some code / design ideas you have?

We have to know what new users find unintuitive, so we can make it more clear or change the API.

**Comment by fdietze** [comment]
Related: #75

**Comment by mariusmuja** [comment]
@SergeantSod In case it helps, you can find here a complete example: https://github.com/mariusmuja/outwatch-extras/blob/master/demo-spa/src/main/scala/demo/DemoApp.scala

In most cases, the code remains almost the same compared to the non-referential transparent API, for example something like this:

```scala
 val names = createStringHandler()

val node = div(
input(placeholder := "Name", inputString --> names),
  hr(),
  h2("Hello ", child <-- names)
)
```

now becomes:
```scala
val node = Handler.create[String].flatMap { names =>
  div(
    input(placeholder := "Name", inputString --> names),
    hr(),
    h2("Hello ", child <-- names)
  )
}
```

As to why `Handler.create[T]` returns an `IO` look for example at line [210](https://github.com/mariusmuja/outwatch-extras/blob/master/demo-spa/src/main/scala/demo/DemoApp.scala#L210) in the example, where there's something like this:

```scala
tr(
  td(TodoModule(loggerSink, store.
  td(TodoModule(loggerSink, store.sink))
)
```

Another way to write it with the current API is:
```scala
val todoModule = TodoModule(loggerSink, store.sink)
tr(
  td(todoModule),
  td(todoModule)
)
```
and it works the same way. With the non-referential transparent API from before, it would display the two todo modules, but they would have the same handler internally, so adding something to one would have resulted in the item being added to both.

**Comment by SergeantSod** [comment]
Thanks for the explanations so far. I think I'll have to peruse the example code @mariusmuja  pointed out a bit more to wrap my head around what it means for my use-case. What I wanted to achieve might already be available as outwatch-extras's `Component`. 

This also touches on @fdietze's question about what I wanted to achieve. My ideas are pretty half-baked, so bear with me here:

I intended to add a thin layer of classes on top of outwatch's API, to allow me to separate views (mostly DOM elements and streams of events) from the other concerns (transformations, extracking data from the model, glue code etc.), which could be pushed into another class (for lack of a better words, let's call it a presenter). For this purpose I envisioned the view class exposing its events along the following lines:

```scala
class ExampleView{
  val clickStream = createInputHandler()

  def element ={
    div(
      "Look at my shiny dom",
      div("More dom",
         clicks --> clickStream
      )
    )
  }
}

```

The intention would be that the presenter would just grab an `ExampleView`'s `clickStream` (and other event streams that are exposed the same way), map over it and pass the resulting stream off to somewhere else. If I extract some traits, I could even reuse different views with the same presenter and vice versa. I would still have to hide the fact that clickStream is also a sink (by narrowing the type) from the outside world, but the idea is the same.

I haven't really figured out how this would work if `clickStream` is an `IO` that will yield a new `Handler` each time it's mapped over. At least the naive approach like the following probably wouldn't work: 

```scala
class ExampleView{
  val clickStream : IO[Handler] = Handler.create[MouseEvent]
  
 def element = {
    for {
       _clickSink <- clickStream
    } yield {
      div(
        "Look at my shiny dom",
        div("More dom",
           onClick --> _clickSink
        )
      )
    }
   
  }
}
```

AFAICS, that way, the presenter could map over `clickStream` again, but that would yield a fresh instance of `Handler[MouseEvent]`, instead of the one that is connected to the DOM element's onClick.
